File: src/lib/domain/research.ts
// Domain types for deep-research agentic search (RFC 0037).
// Shapes mirror the Rust serde (camelCase) in src-tauri/src/domain/research.rs.

import type { DiscoverySearchResponse } from "$lib/domain/discover";

export type Depth = "quick" | "standard" | "thorough";

export type SearchRunStatus =
  | "queued"
  | "planning"
  | "searching"
  | "assessing"
  | "ranking"
  | "ready"
  | "failed"
  | "cancelled";

export interface SearchStrategy {
  depth: Depth;
  maxIterations: number;
  maxProviderQueries: number;
  maxLlmCalls: number;
}

export interface SearchConstraints {
  yearFrom?: number;
  yearTo?: number;
  providers: string[];
  openAccess: boolean;
  targetCount: number;
  venues: string[];
  authors: string[];
  fieldsOfStudy: string[];
  seedPaperIds: string[];
}

export interface SearchDraft {
  title: string;
  goal: string;
  constraints: SearchConstraints;
  strategy: SearchStrategy;
}

export interface Search {
  id: string;
  title: string;
  goal: string;
  constraints: SearchConstraints;
  strategy: SearchStrategy;
  status: string;
  stopReason?: string;
  summary?: string;
  createdAt: string;
  updatedAt: string;
}

export interface SearchRun {
  id: string;
  searchId: string;
  mode: "quick" | "deep" | "improve" | string;
  providerSet?: string;
  queryExpansions?: string;
  status: string;
  stopReason?: string;
  iteration: number;
  addedCount: number;
  totalCount: number;
  providerQueryCount: number;
  llmCallCount: number;
  inspectedCandidateCount: number;
  startedAt?: string;
  finishedAt?: string;
  error?: string;
  createdAt: string;
}

// The candidate payload is the same normalized paper shape that shallow
// discovery returns. Deep research ranks and explains these candidates; it does
// not invent a second paper model.
export type ResearchPaper = DiscoverySearchResponse["candidates"][number];

export interface SearchCandidate {
  id: string;
  searchId: string;
  firstSeenRunId: string;
  rank: number;
  score?: number;
  rationale?: string;
  rankSignalsJson?: string;
  providerHitsJson?: string;
  candidate: ResearchPaper;
  alreadyInLibrary: boolean;
  saved: boolean;
  seen: boolean;
  firstSeenAt: string;
}

export interface SearchUpdated {
  searchId: string;
  runId: string;
  status: string;
  message: string;
  iteration: number;
  found: number;
  unique: number;
  new: number;
}

export interface SearchCandidatesPreview {
  searchId: string;
  runId: string;
  candidates: ResearchPaper[];
  unique: number;
}

export interface HarnessConfiguration {
  goal: string;
  researchInstructions: string;
  scope: string;
  exclusions: string;
  preferredConcepts: string[];
  excludedConcepts: string[];
  sources: string[];
  depth: Depth;
  paperBudget: number;
  autonomy: "manual" | "propose" | "automatic";
  mayAddPapers: boolean;
  writableDocumentIds: string[];
  schedule: HarnessSchedule;
  stopConditions: HarnessStopConditions;
}

export interface HarnessSchedule {
  enabled: boolean;
  cadence: "daily" | "weekly";
  localTime: string;
  weekday?: number;
  timezone: string;
}

export interface HarnessStopConditions {
  maximumCycles?: number;
  endAt?: string;
  maximumUnproductiveRuns?: number;
  maximumRunSeconds?: number;
  maximumProviderQueries?: number;
  maximumLlmCalls?: number;
  stopOnConvergence: boolean;
}

export interface ResearchHarness {
  projectId: string;
  status: string;
  configuration: HarnessConfiguration;
  configurationVersion: number;
  nextRunAt?: string;
  lastScheduledFor?: string;
  requestedPostRunStatus: string;
  completedCycleCount: number;
  consecutiveUnproductiveRuns: number;
  terminalStopReason?: string;
  updatedAt: string;
}

export interface HarnessRun {
  id: string;
  projectId: string;
  status: string;
  configurationSnapshot: HarnessConfiguration;
  configurationVersion: number;
  policyVersion: string;
  effectiveInstructions: EffectiveInstructionStack;
  searchId: string;
  searchRunId?: string;
  startingStateRevision: number;
  resultingStateRevision?: number;
  startingVaultRevision: number;
  resultingVaultRevision?: number;
  providerQueryCount: number;
  llmCallCount: number;
  iterationCount: number;
  inspectedCandidateCount: number;
  trigger: "manual" | "scheduled" | "startup_catch_up";
  scheduledFor?: string;
  stopReason?: string;
  summary?: string;
  startedAt: string;
  finishedAt?: string;
}

export interface HarnessConfigurationVersion {
  projectId: string;
  version: number;
  configuration: HarnessConfiguration;
  actor: "researcher" | "improvement" | "migration" | string;
  sourceImprovementId?: string;
  reason: string;
  createdAt: string;
}

export interface RunContextEntry {
  id: string;
  kind: string;
  epistemicStatus: string;
  text: string;
  lifecycle: string;
}

export interface RunContextObservation {
  kind: string;
  description: string;
}

export interface EffectiveRunContext {
  startingStateRevision: number;
  activeEntries: RunContextEntry[];
  vaultId: string;
  vaultRevision: string;
  vaultPaperIds: string[];
  priorNextDirection?: string;
  priorObservations: RunContextObservation[];
  maximumProviderQueries: number;
  maximumLlmCalls: number;
  paperBudget: number;
}

export interface EffectiveInstructionStack {
  productPolicyVersion: string;
  productPolicySummary: string;
  projectResearchInstructions: string;
  structuredSettings: HarnessConfiguration;
  runContext: EffectiveRunContext;
}

export interface CandidateDecision {
  candidateId: string;
  decision: "accept" | "reject";
  reason: string;
  relevanceConfidence: number;
  withinScope: boolean;
}

export interface PlannedEvidence {
  candidateId: string;
  excerpt: string;
  supportNote?: string;
}

export interface PlannedRelation {
  target: string;
  kind: "derived_from" | "motivated_by";
}

export interface PlannedResearchEntry {
  handle: string;
  kind: "finding" | "question" | "gap" | "hypothesis" | "experiment_idea";
  epistemicStatus: "source_supported" | "agent_synthesis" | "speculative";
  text: string;
  evidence: PlannedEvidence[];
  relations: PlannedRelation[];
}

export interface RunReconciliationPlan {
  candidateDecisions: CandidateDecision[];
  entries: PlannedResearchEntry[];
  nextDirection: string;
  operationalReflection?: PlannedHarnessReflection;
}

export interface PlannedHarnessObservation {
  kind:
    | "query_quality"
    | "irrelevant_result_class"
    | "source_failure"
    | "terminology"
    | "coverage_bias"
    | "relevance_error"
    | "scope_drift"
    | "wasted_work";
  signature: string;
  severity: number;
  confidence: number;
  description: string;
  target?: "preferred_concepts" | "excluded_concepts" | "metadata_resolvers";
  proposedValue?:
    | { kind: "concepts"; items: string[] }
    | { kind: "metadata_resolvers"; items: string[] };
  proposalEligible: boolean;
}

export interface PlannedHarnessReflection {
  summary: string;
  nextDirection?: string;
  observations: PlannedHarnessObservation[];
}

export interface HarnessChangeSet {
  id: string;
  runId: string;
  projectId: string;
  startingStateRevision: number;
  status: "proposed" | "applied" | "rejected" | "superseded" | "failed";
  plan?: RunReconciliationPlan;
  consideredCandidates: SearchCandidate[];
  error?: string;
  decisionReason?: string;
  resultingStateRevision?: number;
  createdAt: string;
  decidedAt?: string;
}

export interface HarnessEvent {
  id: string;
  runId: string;
  sequence: number;
  kind: string;
  summary: string;
  detail?: Record<string, unknown>;
  phase?: string;
  progressCurrent?: number;
  progressTotal?: number;
  actor: "researcher" | "harness" | "scheduler" | "system" | string;
  occurredAt: string;
}

export interface HarnessUsage {
  providerQueries: number;
  llmCalls: number;
  iterations: number;
  inspectedCandidates: number;
}

export interface ResearchCheckpoint {
  runId: string;
  projectId: string;
  status: string;
  startingStateRevision: number;
  resultingStateRevision?: number;
  startingVaultRevision: number;
  resultingVaultRevision?: number;
  appliedChangeSetId?: string;
  acceptedCandidateCount: number;
  rejectedCandidateCount: number;
  addedPaperIds: string[];
  affectedDocuments: Array<{ documentId: string; contentRevision: number }>;
  usage: HarnessUsage;
  stopReason?: string;
  complete: boolean;
  converged: boolean;
  reflectionId?: string;
  nextDirection?: string;
  startedAt: string;
  finishedAt?: string;
  restoreAvailable: boolean;
}

export interface HarnessSnapshot {
  harness: ResearchHarness;
  runs: HarnessRun[];
  events: HarnessEvent[];
}

// Depth presets — mirror Depth::budget() in research.rs.
export function depthStrategy(depth: Depth): SearchStrategy {
  switch (depth) {
    case "quick":
      return { depth, maxIterations: 1, maxProviderQueries: 3, maxLlmCalls: 4 };
    case "thorough":
      return { depth, maxIterations: 4, maxProviderQueries: 16, maxLlmCalls: 24 };
    default:
      return { depth: "standard", maxIterations: 3, maxProviderQueries: 8, maxLlmCalls: 12 };
  }
}

export function isTerminalStatus(status: string): boolean {
  return status === "ready" || status === "failed" || status === "cancelled";
}

File: src-tauri/src/domain/research.rs
//! Domain types for deep-research agentic search (RFC 0037).
//!
//! A `Search` is the durable unit: a saved goal + constraints whose candidate
//! pool *stacks* across runs. A `SearchRun` is one execution (a thin ledger);
//! `SearchCandidate` rows are the stacked pool, each remembering the run that
//! first surfaced it.

// These types are consumed across the RFC 0037 layers (seams, loop, commands,
// frontend serialization); some are not yet wired as the build lands layer by
// layer. Remove once the commands + frontend layers are in.
#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use crate::domain::discovery::{
    paper_candidate_dedup_key, DiscoveryProviderChoice, PaperCandidate,
};

/// How often a saved search re-runs on its schedule.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Frequency {
    Daily,
    Weekly,
    Monthly,
}

/// Schedule configuration for a saved search (Phase 2 automation).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchSchedule {
    pub enabled: bool,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub frequency: Option<Frequency>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub next_run_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub last_run_at: Option<String>,
}

/// Coarse user-facing depth knob; expands to a `SearchStrategy` budget.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Depth {
    Quick,
    Standard,
    Thorough,
}

impl Depth {
    /// Expand the preset into the hard budget rails the loop enforces.
    pub fn budget(self) -> SearchStrategy {
        match self {
            Depth::Quick => SearchStrategy {
                depth: self,
                max_iterations: 1,
                max_provider_queries: 3,
                max_llm_calls: 4,
            },
            Depth::Standard => SearchStrategy {
                depth: self,
                max_iterations: 3,
                max_provider_queries: 8,
                max_llm_calls: 12,
            },
            Depth::Thorough => SearchStrategy {
                depth: self,
                max_iterations: 4,
                max_provider_queries: 16,
                max_llm_calls: 24,
            },
        }
    }
}

/// Hard budget rails for a run, derived from `Depth`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchStrategy {
    pub depth: Depth,
    pub max_iterations: u32,
    pub max_provider_queries: u32,
    pub max_llm_calls: u32,
}

/// Structured constraints for a search. Soft intent stays in `goal`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchConstraints {
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub year_from: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub year_to: Option<i32>,
    #[serde(
        default,
        deserialize_with = "crate::domain::discovery::deserialize_providers_lenient"
    )]
    pub providers: Vec<DiscoveryProviderChoice>,
    #[serde(default)]
    pub open_access: bool,
    pub target_count: i32,
    #[serde(default)]
    pub venues: Vec<String>,
    #[serde(default)]
    pub authors: Vec<String>,
    #[serde(default)]
    pub fields_of_study: Vec<String>,
    #[serde(default)]
    pub seed_paper_ids: Vec<String>,
}

/// Input to create a new search.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchDraft {
    pub title: String,
    pub goal: String,
    pub constraints: SearchConstraints,
    pub strategy: SearchStrategy,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub schedule: Option<SearchSchedule>,
}

/// A saved search (the durable unit).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Search {
    pub id: String,
    pub title: String,
    pub goal: String,
    pub constraints: SearchConstraints,
    pub strategy: SearchStrategy,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub schedule: Option<SearchSchedule>,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub stop_reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub summary: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Lifecycle status of a run (also reused as a search's latest-run status).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SearchRunStatus {
    Queued,
    Planning,
    Searching,
    Assessing,
    Ranking,
    Ready,
    Failed,
    Cancelled,
}

impl SearchRunStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            SearchRunStatus::Queued => "queued",
            SearchRunStatus::Planning => "planning",
            SearchRunStatus::Searching => "searching",
            SearchRunStatus::Assessing => "assessing",
            SearchRunStatus::Ranking => "ranking",
            SearchRunStatus::Ready => "ready",
            SearchRunStatus::Failed => "failed",
            SearchRunStatus::Cancelled => "cancelled",
        }
    }
}

/// One execution of a search — a thin ledger, not candidate storage.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchRun {
    pub id: String,
    pub search_id: String,
    pub mode: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub provider_set: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub query_expansions: Option<String>,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub stop_reason: Option<String>,
    pub iteration: i32,
    pub added_count: i32,
    pub total_count: i32,
    pub provider_query_count: u32,
    pub llm_call_count: u32,
    pub inspected_candidate_count: u32,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub started_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub finished_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub error: Option<String>,
    pub created_at: String,
}

/// A ranked candidate produced by the loop, ready to stack into the pool.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RankedCandidate {
    pub candidate: PaperCandidate,
    pub rank: i32,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub score: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rationale: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rank_signals_json: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub provider_hits_json: Option<String>,
}

/// A persisted candidate in a search's stacked pool.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchCandidate {
    pub id: String,
    pub search_id: String,
    pub first_seen_run_id: String,
    pub rank: i32,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub score: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rationale: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rank_signals_json: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub provider_hits_json: Option<String>,
    pub candidate: PaperCandidate,
    pub already_in_library: bool,
    pub saved: bool,
    pub seen: bool,
    pub first_seen_at: String,
}

/// Normalized dedup key for a candidate: prefer DOI, then arXiv id, then a
/// lowercased/trimmed title. Used by stacking (`diff`) and within-run dedup.
pub fn candidate_dedup_key(candidate: &PaperCandidate) -> String {
    paper_candidate_dedup_key(candidate)
}
